src/asyncrpcqueue.cpp

```diff
@@ -0,0 +1,232 @@
+                break;
+            }
+
+            // Exit if the queue is closing.
+            if (isClosed()) {
+                while (!operation_id_queue_.empty()) {
+                    operation_id_queue_.pop();
+                }
+                break;
+            }
+
+            // Get operation id
+            key = operation_id_queue_.front();
+            operation_id_queue_.pop();
+
+            // Search operation map
+            AsyncRPCOperationMap::const_iterator iter = operation_map_.find(key);
+            if (iter != operation_map_.end()) {
+                operation = iter->second;
+            }
+        }
+
+        if (!operation) {
+            // cannot find operation in map, may have been removed
+        } else if (operation->isCancelled()) {
+            // skip cancelled operation
+        } else {
+            operation->main();
+        }
+    }
+}
+
+
+/**
+ * Add shared_ptr to operation.
+ *
+ * To retain polymorphic behaviour, i.e. main() method of derived classes is invoked,
+ * caller should create the shared_ptr like this:
+ *
+ * std::shared_ptr<AsyncRPCOperation> ptr(new MyCustomAsyncRPCOperation(params));
+ *
+ * Don't use std::make_shared<AsyncRPCOperation>().
+ */
+void AsyncRPCQueue::addOperation(const std::shared_ptr<AsyncRPCOperation> &ptrOperation) {
+
+    // Don't add if queue is closed or finishing
+    if (isClosed() || isFinishing()) {
+        return;
+    }
+
+    AsyncRPCOperationId id = ptrOperation->getId();
+    std::lock_guard< std::mutex > guard(lock_);
+    operation_map_.emplace(id, ptrOperation);
+    operation_id_queue_.push(id);
+    this->condition_.notify_one();
+}
+
+/**
+ * Return the operation for a given operation id.
+ */
+std::shared_ptr<AsyncRPCOperation> AsyncRPCQueue::getOperationForId(AsyncRPCOperationId id) const {
+    std::shared_ptr<AsyncRPCOperation> ptr;
+
+    std::lock_guard< std::mutex > guard(lock_);
+    AsyncRPCOperationMap::const_iterator iter = operation_map_.find(id);
+    if (iter != operation_map_.end()) {
+        ptr = iter->second;
+    }
+    return ptr;
+}
+
+/**
+ * Return the operation for a given operation id and then remove the operation from internal storage.
+ */
+std::shared_ptr<AsyncRPCOperation> AsyncRPCQueue::popOperationForId(AsyncRPCOperationId id) {
+    std::shared_ptr<AsyncRPCOperation> ptr = getOperationForId(id);
+    if (ptr) {
+        std::lock_guard< std::mutex > guard(lock_);
+        // Note: if the id still exists in the operationIdQueue, when it gets processed by a worker
+        // there will no operation in the map to execute, so nothing will happen.
+        operation_map_.erase(id);
+    }
+    return ptr;
+}
+
+/**
+ * Return true if the queue is closed to new operations.
+ */
+bool AsyncRPCQueue::isClosed() const {
+    return closed_.load();
+}
+
+/**
+ * Close the queue and cancel all existing operations
+ */
+void AsyncRPCQueue::close() {
+    closed_.store(true);
+    cancelAllOperations();
```

Comment:
I think this is not guaranteed to ensure that all operations are cancelled when it returns. The reason is that the check for ``isClosed() || isFinishing()`` in ``addOperation`` is done outside the lock. So it is possible for ``addOperation`` to get past that check, and then ``close()`` to be called in a different thread and grab the lock (in ``cancelAllOperations()``) before ``addOperation`` does, and then when it completes, ``addOperation`` unblocks and adds to the queue.

It would be more obviously correct to not use an atomic for ``closed_`` and just ensure that the lock is held for all accesses to it.